Feature: Provenance

benloh opened the pull request:
This adds support for provenance:
* Any field can be added to the "Provenance" section of Node and Edge editors by adding the `isProvenance` flag.
* Provenance fields honor the `hidden` property as any non-built-in field does (e.g. can be hidden, and will not be exported when hidden).
* Introduces a new field `infoOrigin` that initially displays `Created by <createdBy> on <created>` but then operates as string field.  It can be used outside of the "Provenance" section if you want.
* The "Created" and "Updated" history rows will be hidden if `created` or `updated` are hidden (respectively), but will still be exported and imported because they are built-in fields.
* `createdBy` and `updatedBy` can be selectively hidden so that the History will hide author info and only show "Created: 1/7/2025, 10:05:37 AM" or "Updated: 1/7/2025, 10:05:37 AM".

# TBD
* Displays provenance in tables
* Can be filtered


```
Provenance Tab

Provenance Section
* SOURCE: Created by BEN-QPIX on May 10, 10:10am

History Section
* Created: DEMO-A-JW7, 1/4/2025, 10:34:58 AM
* Updated: DEMO-A-JW7, 1/4/2025, 10:35:00 AM
* Revision: 2
```

See also #55, #275, #276



# Wiki 
(To be added with PR)


## `infoOrigin` field

`infoOrigin` is a special field.  Initially the value will be set to `Created by <createdby> on <created>`, but upon editing it behaves like a string field so you can edit the field to say anything you want.  Once you edit the field, the field behaves like a string field.

`infoOrigin` honors both `isProvenance` and `hidden` parameters.


## Provenance Tab


### Adding fields to the "Provenance" section

In project templates, use `isProvenance = true` to add a field to the Provenance Section of a Node or Edge Editor.  Any field type can be supported.


### "History" section

The "History" section of the tab uses five built-in fields:
* `created` -- date of creation
* `createdby` -- token of user who created the object
* `updated` -- date of update of the object
* `updatedby` -- token of user who updated the object
* `revision` -- a count of the number of times the object has been revised.

The "History" shows a combination of fields:
* "Created" = `createdby` + `created`
* "Updated" = `upatedby` + `updated`

#### `hidden` parameter
The history fields may be hidden in the Node/Edge Editors and Tables by using `hidden = true`, but they will always be exported and imported.

* If `created` is hidden, then the whole "Created: <token>, <date>" line is hidden.
* If `createdby` is hidden, then only the user is hidden, e.g. the line will display "Created: <date>"
* `updated` and `updatedby` work the same way.

#### `isProvenance` parameter
The "History" section ignores the `isProvenance = true` flag.  It is a built-in field that will always be placed in the "History" section (as long as it's not hidden).




## `timestamp` field filtering

Since we do not have a specialized UI for setting time and date (e.g. you can't select a particular day or year or time), we rely on the Historical Date functionality to provide a UI for specifying a filter string.  e.g. you can search by using:
```
> 2024 ad
> 2024 ad 4pm
> jan 1
> 12pm
> jan 1 2025
```

benloh committed:
match-meme-comment: Show comment type

Match MEME

benloh committed:
match-meme-comment: Show simple "DELETED" for comment prompts

Rather than show complex "DELETED" for every prompt in a comment type because there may be many.

benloh committed:
match-meme-comment: Fix bug: `commenter_id` is inadvertently changed when updating comment status.

benloh committed:
match-meme-comment: Count only non-deleted comments.  Add `commentCount` to TCommentCollection

benloh committed:
match-meme-comment: Ignore deleted comments in comment stats (unread, repliesToMe)

benloh committed:
match-meme-comment: Backport URTable

benloh committed:
Merge branch 'dev-next' into dev-bl/match-meme-comment

# Conflicts:
#	app/view/netcreate/components/NCEdgeTable.jsx
#	app/view/netcreate/components/NCNodeTable.jsx

benloh committed:
Merge branch 'dev-next' into dev-bl/match-meme-comment -- missing NCTemplate.jsx

benloh committed:
fix-export: Fail gracefully if node type options or edge type options have not been defined.

benloh committed:
match-meme-comment: Match MEME URCommentPrompt

benloh committed:
match-meme-comment: Match MEME URCommentStatus with UNISYS fixes

benloh committed:
match-meme-comment: Match MEME URCommentThread

benloh committed:
match-meme-comment: Match MEME URComment with UNISYS fix

benloh committed:
match-meme-comment: Match MEME ac-comments

benloh committed:
match-meme-comment: Match MEME dc-comments

benloh committed:
match-meme-comment: Port URCommentSVGBtn and URCommentCollectionMgr with UNISYS fixes

benloh committed:
match-meme-comment: Match URCommentVBtn with UNISYS fixes

benloh committed:
match-meme-comment: Add svg comment icon to comment-mgr (replace svgjs implementation in MEME)

benloh committed:
match-meme-comment: Match some of MEME comment-mgr calls

benloh committed:
match-meme-comment: Update URCommentVBtn when cref changes so count is updated when Node selection is switched

benloh committed:
match-meme-comment: Match MEME add URDialog from URCommentStatus

benloh committed:
match-meme-comment: Match MEME replace URCommentThreadMgr with URCommentCollectionMgr

benloh committed:
match-meme-comment: Convert NCNode, NCEdge, NCNodeTable, NCEdgeTable to use URCommentVBtn (and URCommentCollectionMgr)

benloh committed:
match-meme-comment: Match MEME URComment css

benloh committed:
match-meme-comment: NC uses `comment.comment_id` where MEME uses `comment.id` to match pmcData

benloh committed:
match-meme-comment: Match method name casing

benloh committed:
match-meme-comment: Match MEME comment-mgr calls, especially support for URCommentCollectionMgr calls.

benloh committed:
match-meme-comment: Deprecate toggle comment button.  Currently handled directly by URCommentVBtn but we might want to restore this

to force close the currently open comment window and re-open the table comment button

benloh committed:
match-meme-comment: Refactor NCNodeTable and NCEdgeTable as functional components.  Simplify URDATA handlers."

"SELECTION" and "HILITE" are not yet implemented.

benloh committed:
match-meme-comment: Comment types are defined first in project template, then dc-comments.DEFAULT_CommentTypes then fall back to DEFAULT_COMMENTTYPE #244

benloh committed:
match-meme-comment: project template comment types should override any defaults defined in dc-comments

benloh committed:
match-meme-comment: Tighten up comment spacing, add more gap between prompts.

benloh committed:
lint cleanup debug

benloh committed:
match-meme-comment: Wait for comment to open before trying to scroll to it.  Fixes #321

benloh committed:
lint rewrap

benloh committed:
provenance: Honor `hidden=true` for created, updated, revision in editors

benloh committed:
provenance: Remove `isProvenance` from built in fields for _default.tempate.toml

`isProvenance` is irrelevant for built-in fields, so make it clearer that it has no effect

benloh committed:
provenance: Add custom `infoOrigin` field that initially displays `Created by <created> on <createdBy>` #55

benloh commented:
@jdanish Basic support for Provenance and `infoOrigin` is now available.  Tables and filters are not yet operational.

* Checkout `dev-bl/provenance`
- [x] Try adding a `infoOrigin` field type to nodes:
```
[nodeDefs.origin]               // you can also call this 'nodeDefs.source' or 'nodeDefs.citation' etc 
type = "infoOrigin"             // this would be the name of the specialized type
displayLabel = "Information Origin"
exportLabel = "InfoOrigin"
help = "Where did the information come from?"
includeInGraphTooltip = true
isProvenance = true
hidden = false
```
- [x] Try adding a `infoOrigin` field type to edges.
- [x] Try moving a field in and out of "Provenance" by changing the `isProvenance` parameters.
- [x] Try hiding and showing the `infoOrigin` field by changing the `hidden` parameter -- it should be hidden and NOT exported since it's not a built-in field.
- [x] Try hiding and showing the "Created" row by changing the `hidden` parameter for the built-in field: `created` -- it should not be displayed, but still exported.
- [x] Try hiding and showing the "Updated" row by changing the `hidden` parameter for the built-in field: `updated` -- it should not be displayed, but still exported.
- [x] Try adding new node and make sure infoOrigin information is correctly added
- [x] Try adding new edge and make sure infoOrigin information is correctly added

Update: Joshua added last 2 bullets

jdanish commented:
Before I test - should we use old files, new files, doesn't matter?

jdanish commented:
Nevermind, tried with existing file and simple changed InfoSource attributes to type infoOrigin and it worked.

jdanish commented:
@benloh confirmed this works but that the contents of the infoSource field only export if I have changed them. The default text appears in the attribute editing panel, but not in table or export. I assume that is part of the "this does not work in tables" yet, so am not reporting a separate issue, though I can if you need. Otherwise all the steps seem right and it's great.

benloh commented:
Ah, that's a good point.  I was trying to avoid a situation where we have to create a new initialization phase to predefine fields but that's probably wishful thinking.  I'll get that added.

jdanish commented:
Well, the funny thing is that it is displaying in the attribute panel but presumably not saving that value, so maybe just tweak that code to say if you are displaying and there was nothing there, go ahead and save it? Then whatever displays it first will fill it (panel, table, etc).

benloh committed:
lint rewrap

benloh committed:
provenance: Move `DeriveInfoOriginString` to `nc-utils` so importexport-mgr can use it as well

benloh committed:
provenance: Generate `infoOrigin` during exports. #322

benloh commented:
We want to keep the data operations of creating a new node/edge in the database and creating a new node/edge in the UI agnostic with regard to custom fields so that node and edge data remain "pure". So we're introducing special handling for `infoOrigin` fields that are only activated when a node or edge is displayed in the editor or table or is exported.

[c6d9873](https://github.com/netcreateorg/netcreate-itest/pull/322/commits/c6d98737828d682ab1dd2f4b5583c783af6c239b) and [7efd86a](https://github.com/netcreateorg/netcreate-itest/pull/322/commits/7efd86a6cc0c92922a98a070b479d5aebd566f7c) should implement that.

Notes from `nc-ui.js`
```
Custom Fields Handling

    The `infoOrigin` field requires special handling.

    Ideally, when a node is created, we would process any special fields and
    derive and save any derived info.  But
    a) we don't want to have server-database handle any custom fields -- that
       it should just blindly save data.
    b) when nc-logic creates a new field via `NODE_CREATE`, again, only built-in
       fields are created and stored.  Custom fields are created by the UI as
       needed when they are rendered.
    This means that the `infoOrigin` field remains blank until the user
    decides to change the data.

    To handle this, then we need to make sure `infoOrigin` appears to have data
    whenever it is viewed.  This requires:
    - nc-ui renders a non-built-in field, e.g. an `attribute` field
    - nc-ui renders a `provenance` field
    - importexport-mgr exports node or edge data

```

jdanish commented:
Confirmed fixed in export, but not yet in tables.

benloh commented:
> but not yet in tables.

Yes, I'm working on tables.  It's much more complicated because we have to deal with the space issues.

jdanish commented:
Cool - I figured but wanted to make sure it was all explicit.

benloh committed:
provenance: Fix _default.template.toml so "Label" is capitalized, and edge "info/Date" doesn't refer to mismatched date/number

benloh committed:
provenance: URTable markdown sort show empty rows at bottom

benloh committed:
provenance: Limit height of images in tables.  CSS should be in URTable.css not NCNode.css.

benloh committed:
provenance: Show Provenance in Node and Edge Tables.

benloh committed:
provenance: Fail gracefully if provenance data is not yet defined

benloh committed:
provenance: Add support for `hdate-short` and `timestamp-short` column definitions.

benloh committed:
provenance: Add `infoOrigin` support to NCNodeTable and NCEdgeTable.

benloh committed:
provenance: Use "" for Comment column header.  #325

benloh committed:
provenance: Don't show tabletips on View/Edit and Comment buttons #326

benloh committed:
provenance: Use template labels for history columns. #327

benloh committed:
provenance: Show table popups in-line #324

benloh committed:
lint rewrap

benloh committed:
provenance: Add rudimentary support for `infoOrigin` and `timestamp` filter types

benloh committed:
provenance: Add support for filtering timestamps (which are not Historical Dates)

benloh committed:
provenance: Restore hdate view in tables

benloh committed:
provenance: Tables need to use `meta.created` when deriving `infoOrigin`

benloh committed:
provenance: Style date fields to match other Node/Edge editor fields. (indent)

benloh committed:
match-meme-comment: Match MEME backport of Comment Cancel. Fixes #329

benloh committed:
provenance: Catch missing `meta` parameters with new nodes and edges. Fixes Edge table breaking upon creating a new edge.

benloh committed:
provenance: Clean up `revision` for older projects so that canceling an Edge Edit will not delete the edge.  #330.

benloh committed:
provenance: Optimize Node and Edge editor rendering

benloh committed:
provenance: Fix missing setState

benloh committed:
provenance: Show larger images on table tip hover. #333

benloh committed:
provenance: Case insensitive sorting for node labels in tables.  Use template-defined sort types where appropriate, but also override for styling.

created and updated use `timestamp-short`.  createdBy and updatedBy use `text-case-insensitive`

benloh committed:
provenance: Fix duplicate key warning in provenance tabs

benloh commented:
For the record...this branch/PR was successfully used with the Jan 15 IU pilot.
